[WIP] Initial attempt at adding lights.

simse opened the pull request:
## Description:
Started adding lights. *However, I need help* The light isn't showing up in the home app. I am aware that it wouldn't work, but I would assume that it would at least show up. What am I missing here? Also, sorry for the messy code, I was trying to debug.

## Checklist:
  - [ ] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

simse committed:
Initial attempt at adding lights.

simse committed:
Initial attempt at adding lights.

simse committed:
Initial attempt at adding lights.

homeassistant commented:
Hi @simse, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#12576).

Once you do that we will be able to review and accept this pull request.

Thanks!

Kane610 commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -68,7 +68,9 @@ def import_types():
-    from .sensors import TemperatureSensor # noqa F401
+    from .sensors import TemperatureSensor  # noqa F401
+    # pylint: disable=unused-variable
+    from .lights import Lightbulb  # noqa F401
```
I think lightbulb is to specific since it would also represent a led strip among other things

simse replied:
Will bear that in mind.

cdce8p replied:
"Lightbulb" is just the name of the "service" assigned by Apple. This shouldn't be the main concern right now.

cdce8p commented:
@simse What is your Log output? Do you see the debug statements? I tried your code with a template light and it seems to work just fine. Maybe consider removing the "Home Assistant" Bridge form "Home" and adding it again.

simse commented:
@cdce8p First of all, it's great to hear it's working for you. I'm not even seeing a bridge, I'll try running it on my Macbook instead. My log output is:
```
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit] Import type files.
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit] Start adding accessories.
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component1
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component2
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component3
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component4
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component5
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component6
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component7
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component8
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component9
2018-02-21 16:04:29 DEBUG (Thread-10) [homeassistant.components.homekit.lights] Loading light component10
```

So I am assuming the component was loaded?

cdce8p commented:
Yes it should work. You can access the bridge from the details panel of the accessories or from the default room (at least in my case).

simse commented:
@cdce8p Running on my Mac works fine. I was running on Ubuntu subsystem on Windows 10 before, and that is likely what caused the problem.

cdce8p commented:
@simse Good to hear. I'm not sure though, since I'm developing from Ubuntu as well ;)

simse commented:
@cdce8p Are you using Ubuntu on Windows 10 as well, or Ubuntu OS native?

cdce8p commented:
Win 7 running VirtualBox with an Ubuntu installation.

cdce8p commented on homeassistant/components/homekit/lights.py:
```diff
@@ -0,0 +1,72 @@
+        _LOGGER.debug("Loading light component2")
+        self.set_accessory_info(entity_id)
+        _LOGGER.debug("Loading light component3")
+        self.add_preload_service(SERVICES_LIGHTBULB)
```
To save you some headache later. It seems that `add_preload_service` doesn't work if you plan to add an additional characteristic to the service. This has to be done before the service is added to the accessory. I don't know the reason to it jet. This could work:
```py
from pyhap.loader import get_serv_loader
from pyhap.loader import get_char_loader
serv = get_serv_loader().get(SERVICES_LIGHTBULB)
brightness = get_char_loader().get(CHAR_BRIGHTNESS)
serv.add_opt_characteristic(brightness)
self.add_service(serv)
```

simse replied:
So that's why that wasn't working.. Thank you!

simse committed:
Now supports basic on/off functionality of lights.

simse committed:
Now supports basic on/off functionality of lights.

simse committed:
Now supports basic on/off functionality of lights.

cdce8p commented on homeassistant/components/homekit/lights.py:
```diff
@@ -0,0 +1,73 @@
+        self._hass.services.call(
+            'light', action,
+            {'entity_id': self._entity_id}
+        )
```
Try implementing a test weather or not the `set_value` call came from HA or Homekit. Unfortunately the `setter_callback` is called in both scenarios in the current version of `HAP-python`, which results in at least to service calls (one from HomeKit, the other form the state change listener).

A solution could be a flag variable which is set during `update_state` and removed during `set_state`.

simse replied:
Yeah good idea, my lights were acting very strange.

amelchio committed:
Optimize logbook SQL query (#12608)

swbradshaw committed:
OpenGarage - correctly handle offline status (#12612) (#12613)

balloob committed:
Fix voluptuous breaking change things (#12611)

* Fix voluptuous breaking change things

* Change xiaomi aqara back

amelchio committed:
Move recorder query out of event loop (#12615)

bakedraccoon committed:
Add Tahoma scenes (#12498)

* add scenes to platform

* add scene.tahoma

* requires tahoma-api 0.0.12

* update requirements_all.txt

* hound

* fix pylint error

OttoWinter committed:
Cast unique_id and async discovery (#12474)

* Cast unique_id and async discovery

* Lazily load chromecasts

* Lint

* Fixes & Improvements

* Fixes

* Improve disconnects

cast.disconnect with blocking=False does **not** do I/O; it simply sets an event for the socket client looper

* Add tests

* Remove unnecessary calls

* Lint

* Fix use of hass object

syssi committed:
The name of the enum must be used here because of the speed_list. (#12625)

The fan.set_speed example value is lower-case and led to confusion. Both spellings are possible now: Idle & idle

pvizeli committed:
Hassio update timeout filter list (#12617)

* Update timeout filter list

* Update http.py

kennedyshead committed:
Fix mclimate accounts with not only melissa components (#12427)

* Fixes for mclimate accounts with not only melissa components

* Fixes melissa sensor to only use HVAC

* Bumping version to 1.0.3 and remove OP_MODE that is not supported

* Removes STATE_AUTO from translation and tests

OttoWinter committed:
Fix cast doing I/O in event loop (#12632)

balloob committed:
Remove braviatv_psk (#12669)

balloob committed:
Merge pull request #12609 from home-assistant/release-0-64

0.64.0

balloob committed:
Version bump to 0.64

cdce8p commented:
@simse How are you getting along? Can I help with something?

I commited changes to the `homekit` component yesterday #12676 which should make it easier for you:
* Updated pyhap to v1.1.7 --> Use `char.set_value(value, should_callback=False)` when you want to push changes from HA. This should eliminate the need to check who called the `setter_callback` method, since then it should always be `homekit`.
* Changes in `accessories.py`:
  * `super().__init()` now takes `display_name`, `model` and `category` as parameter
  * `add_preload_service` allows additional chars to be added to a service, like `brightness`

To use them, rebase your branch against `upstream/dev`

simse commented:
@cdce8p That sounds great. I'll probably submit light support within the end of the week.

cdce8p commented:
Looking forward to it. Let me know if I can help.

simse committed:
Merge branch 'dev' of https://github.com/simse/home-assistant into dev

simse committed:
Merge branch 'dev' of https://github.com/simse/home-assistant into dev
